### ISSUE TITLE
support custom cue settings

### ISSUE THREAD

**Issue by alexdean** [w3c/webvtt#286]
I would like the WebVTT spec to define a method for publishers to include custom cue settings.

I would like to be able to mark some cues as the beginning of a paragraph. At [TED Conferences](http://www.ted.com), we use data like this to construct a readable transcript from the subtitles created by [our volunteer translators](http://www.ted.com/about/programs-initiatives/ted-open-translation-project). You can see an example in [the English transcript of Shonda Rhimes'
TEDTalk](http://www.ted.com/talks/shonda_rhimes_my_year_of_saying_yes_to_everything/transcript?language=en).

We currently use a custom JSON format to transmit this data. We would like to switch to a more standard format, but need support for this additional per-caption metadata.

Something like [HTML's data-* attributes](https://www.w3.org/TR/2011/WD-html5-20110525/elements.html#embedding-custom-non-visible-data-with-the-data-attributes) would serve our purposes nicely.

For example:

```
WEBVTT

00:00:00.960 --> 00:00:04.616 data-paragraph:true
So a while ago, I tried an experiment.

00:00:04.640 --> 00:00:08.080
For one year, I would say yes
to all the things that scared me.
```

cc @zcorpan @jwarchol @bendk

**Comment by silviapfeiffer** [comment]
What stops you from using it now? You would like these to not be thrown out by a conformance checker?

**Comment by alexdean** [comment]
@silviapfeiffer Yes, that. But also I would like to be confident that the WebVTT we create will not cause problems for standards-compliant parsers.

**Comment by silviapfeiffer** [comment]
Have you thought of using comments to achieve this? https://w3c.github.io/webvtt/#introduction-comments

Something like

``
WEBVTT

NOTE data-paragraph:true

00:00:00.960 --> 00:00:04.616
So a while ago, I tried an experiment.

00:00:04.640 --> 00:00:08.080
For one year, I would say yes
to all the things that scared me.
``

How do you intend to use this? By parsing the original WebVTT file and creating the HTML from that or by using the VTTCue DOM API?

**Comment by alexdean** [comment]
With comments, I was concerned by [this statement](https://w3c.github.io/webvtt/#webvtt-comment-block), which made me think a parser is expected/required to discard comments.

> A WebVTT comment block is ignored by the parser

Unless this directive were changed to require parsers to preserve comments, the parsing step would end up discarding important information from our documents.

Additional issue: To be suitable for our purposes, a parser algorithm would have to explicitly attach comments to the next cue which follows a comment. This seems like an unsafe assumption. For example, a document could contain both comments describing the entire document and a comment describing a single cue. How will a parser decide which is which?

```
WEBVTT

NOTE The transcriber was Alex Dean.
NOTE The reviewer was Josh Warchol.

00:00:00.960 --> 00:00:04.616
So a while ago, I tried an experiment.

NOTE data-paragraph:true

00:00:04.640 --> 00:00:08.080
For one year, I would say yes
to all the things that scared me.
```

To your second question: We will parse the original WebVTT file and create HTML from that. I'm not familiar with the VTTCue DOM API.

**Comment by dwsinger** [comment]
Perhaps we need to say that there is a difference between the formal parser described in the specification, and the actual parsers in practice.  Here “discard” means “has no effect on further processing or output for reference players”, or something like, not a prohibition on e.g. printing the comments on T-shirts and giving them away to viewers…


> On Feb 21, 2016, at 9:53 , Alex Dean <notifications@github.com> wrote:
> 
> With comments, I was concerned by this statement, which made me think a parser is expected/required to discard comments.
> 
> A WebVTT comment block is ignored by the parser
> 
> Unless this directive were changed to require parsers to preserve comments, the parsing step would end up discarding important information from our documents.
> 
> Additional issue: To be suitable for our purposes, a parser algorithm would have to explicitly attach comments to the next cue which follows a comment. This seems like an unsafe assumption. For example, a document could contain both comments describing the entire document and a comment describing a single cue. How will a parser decide which is which?
> 
> WEBVTT
> 
> NOTE The transcriber was Alex Dean.
> NOTE The reviewer was Josh Warchol.
> 
> 00:00:00.960 --> 00:00:04.616
> So a while ago, I tried an experiment.
> 
> NOTE data-paragraph:true
> 
> 00:00:04.640 --> 00:00:08.080
> For one year, I would say yes
> to all the things that scared me.
> 
> To your second question: We will parse the original WebVTT file and create HTML from that. I'm not familiar with the VTTCue DOM API.
> 
> —
> Reply to this email directly or view it on GitHub.
> 

David Singer
Manager, Software Standards, Apple Inc.

**Comment by zcorpan** [comment]
Some things come to mind:

* Comments and invalid settings will be dropped during parsing, so it's not particularly robust that they will survive e.g. an editing session. Although I would hope that editors don't drop comments, at least. (And the spec allows editors to preserve invalid-anything.)
* Comments and invalid settings are not represented in any way in the DOM API. So it can't be accessed in JS without fetching the file again. It looks like you don't generate the transcript on the client with JS, so this shouldn't matter to you, but it might matter for other applications.

> For example, a document could contain both comments describing the entire document and a comment describing a single cue. How will a parser decide which is which?

Well, that's for you to decide. It seems like a non-problem to me; you could just make the ones saying "NOTE new paragraph" apply to the next cue, and the per-file comments you recognise apply to the file, and ignore any comments you don't recognise.

Another option would be if we introduce cue-level classes, as suggested in https://github.com/w3c/webvtt/issues/272 . I think classes should cover this use case nicely and be able to represent the same information as custom settings would, and has the benefit of being simpler to select with CSS.

**Comment by zcorpan** [comment]
Also note that you can already use classes in the cue text, like:

```
WEBVTT

NOTE The transcriber was Alex Dean.
NOTE The reviewer was Josh Warchol.

00:00:00.960 --> 00:00:04.616
<c.paragraph>So a while ago, I tried an experiment.</c>

00:00:04.640 --> 00:00:08.080
<c.paragraph>For one year, I would say yes
to all the things that scared me.</c>
```

**Comment by silviapfeiffer** [comment]
@zcorpan are you thinking of classes being allowed to be a random string? So, something like this should just work?

```
WEBVTT

00:00:00.960 --> 00:00:04.616 new_paragraph
So a while ago, I tried an experiment.

00:00:04.640 --> 00:00:08.080
For one year, I would say yes
to all the things that scared me.
```

**Comment by zcorpan** [comment]
Yeah, anything except whitespace and `.` (we don't even need to ban `-->` since it appears on the same line as the actual `-->`). Though my suggested syntax would be

```
00:00:00.960 --> 00:00:04.616.new_paragraph
```

**Comment by silviapfeiffer** [comment]
I'd recommend requiring a space in between, unless you want to require classes before cue settings.

**Comment by zcorpan** [comment]
Classes before cue settings, yes. But could require a space anyway (for document conformance at least).

**Comment by nigelmegitt** [comment]
Sorry am a bit late to this, but one option you might want to consider as an alternative for expressing this data in a structured form is to start with [TTML](http://www.w3.org/TR/ttml1/): it already supports semantic constructs like div and p and allows metadata roles, speaker ids etc to be assigned to content elements, for example. You could use an xml to json converter if that makes it easier for you to process client side. It's just a thought: I'm not trying to be divisive, just responding to the way you structured the question @alexdean . No problem also addressing this use case in WebVTT.

**Comment by bendk** [comment]
Couple of opinions:

- The ideal situation, as I understand it, is for parsers to handle the data in some way.  Clearly anyone could write their own parser and deal with comments/invalid cue settings in whatever way they want.  However:
  - It's much easier to use an existing parser and have the parser handle extraction/lookup of the data.
  - You have the issue of other parsers/editors modifying the data as zcorpan points out.
- I think comment blocks could be work, but like Alex points out, if you want to use a standard parser to test for the presence of the comment, then you need to put something in the standard about how comments are linked to cues.  I'm not sure exactly how you would do that since comments are allowed basically anywhere in the file.
- Classes could work, although semantically I don't know if it makes sense.  The paragraph logically wraps all of the cues, not the one that ends the paragraph.  I feel like any solution like this is somewhat of a hack.  It would be nice if you could have a `<c>` tag that spans multiple cues, but I'm guessing there would be other issues with that one.  Another option would be to have an empty tag like `<c.paragraph-break></c>`.  I think that's somewhat more logical, but still seems somewhat hacky to me.  Lastly, it's a bit more difficult on the code side to search for tags in the cue text than to test for a cue setting being present.
  - That said, if cue-level classes are something with momentum behind them, then I think they would satisfy the requirements well enough.  They also seem easy to test for.
- TTML handles this great, but VTT is nicer in many ways and we want to use it.

**Comment by alexdean** [comment]
@nigelmegitt we've definitely considered TTML/DFXP, and it still may be an option for us. WebVTT is preferable because it is much simpler and seems to have nearly everything we need. Additionally, we serve the majority of our video via HLS (which requires subtitles in WebVTT).

Cue-level classes seem like they would address our immediate need. What is the timeline for making a decision on the adopting or rejecting that idea?

**Comment by dwsinger** [comment]
> On Feb 22, 2016, at 8:40 , Alex Dean <notifications@github.com> wrote:
> 
> @nigelmegitt we've definitely considered TTML/DFXP, and it still may be an option for us. WebVTT is preferable because it is much simpler and seems to have nearly everything we need. Additionally, we serve the majority of our video via HLS (which requires subtitles in WebVTT).
> 
> Cue-level classes seem like they would address our immediate need. What is the timeline for making a decision on the adopting or rejecting that idea?

There is no a priori reason we shouldn’t adopt something once it’s thought-through, and proposed.  However, we would dearly love to go to CR really soon, so we’d need to decide whether this would be in that, or in an editor’s darft post-CR.

David Singer
Manager, Software Standards, Apple Inc.

**Comment by silviapfeiffer** [comment]
FAICT cue classes would we rather simple to specify and implement, unless I overlooked any complexities. I suggest we go ahead with it.

**Comment by alexdean** [comment]
i renamed the issue to better describe the solution which seems to have most support.

@zcorpan i see you added the 'needs implementer interest' tag. can you outline what needs to be done? (i'd like to do a PR but i'm not sure where to start.)

**Comment by zcorpan** [comment]
I'd like stated interest to implement this in a browser engine before putting this in the spec. If we go ahead and put things in the spec that won't be implemented in browsers, the spec is not so useful. cc @rgov 

I'm happy to guide you if you want to do a PR. You can use https://api.csswg.org/bikeshed/ to generate (or install bikeshed locally), and ask me in [\#whatwg](https://wiki.whatwg.org/wiki/IRC) on IRC if you have questions. Adding classes to cues touches several parts of the spec; syntax, parsing, selectors, maybe the DOM mapping, maybe the DOM API also.
